### ISSUE TITLE
Stop the gradient computation for Keras eager execution mode

### ISSUE THREAD

**Issue by nairouz** [tensorflow/tensorflow#22378]
------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Colab
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**: 
- **TensorFlow version (use command below)**: 1.10.1
- **Python version**: python 3.6
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

### Describe the problem
I  am facing problems when implementing some custom loss function in Keras based on the eager execution mode. The problem is that some loss functions require to stop the gradient computation for some specific variables. When executed in a graph, we can use the op tf.stop_gradient. In fact, for Keras the GradientTape is internally handled by the function [`_process_single_batch` ](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training_eager.py)

### Source code / logs
Here is an example : 
   
```
    def virtual_adversarial_loss(X, DAE_encoder):
        r_vadv = generate_virtual_adversarial_perturbation(X, DAE_encoder)
        tape.reset()
        tape.watch(X)
        tape.watch(r_vadv)
        p =  DAE_encoder(X)
        p = tape.watch(p)
        q = DAE_encoder(x+r_vadv)
        loss = kl(p, q)
        return tf.identity(loss, name="vat_loss")
```

```
    def virtual_adversarial_loss(X, DAE_encoder):
        r_vadv = generate_virtual_adversarial_perturbation(X, DAE_encoder)
        tf.stop_gradient(X)
        tf.stop_gradient(r_vadv)
        p =  DAE_encoder(X)
        p = tf.stop_gradient(p)
        q = DAE_encoder(x+r_vadv)
        loss = kl(p, q)
        return tf.identity(loss, name="vat_loss")
```

Both of these functions do not work. The problem with the first one is that I can't use the tape variable created by the internal Keras wrapper function [`_process_single_batch` ](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training_eager.py). The problem with the second one is that I can't use tf.stop_gradient for the eager execution mode.

Is there a way to stop the gradient computation in Keras eager execution mode?

**Comment by wt-huang** [comment]
@nairouz just ran above snippets in eager execution and both of these functions should work. You would need to add something like:
`from tensorflow import keras
from tensorflow.python.eager.backprop import GradientTape
from tensorflow import keras as ks
`
Make sure to add `with GradientTape() as tape:` in your first function.

**Comment by nairouz** [comment]
@wt-huang thank you for the response. Yes, I made sure to add all the required imports.

Please let further explain the problem. Some loss functions require to stop the gradient for some specific variables. In fact, there are two cases:

1) In the graph execution mode, one can use the  `tf.stop_gradient(input=a)` and this operator will exclude the "a" variable from any gradient backpropagation.

2) In the eager execution mode, a tape should be created based on the GradientTape class to record the operations since we do not have a graph and only then we can exclude some variables from the gradient backpropagation.

Unfortunately in Keras, one should not create a tape himself since the gradient computation will be done using the tape created by tensorflow developers in this function [`_process_single_batch`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training_eager.py#L479). Even if I create a tape myself, it will not be used for the gradient computation performed by Keras engine unless I implement my own training loop function and hence have my own  backpropagation. In this case,  I would be deprived of Keras high-level API functions like `train` and `predict`.

I think it will be quite useful to extend tensorflow with the following operator:
tf.get_GradientTape() which can return the actual gradient context manager if there is one.

**Comment by nairouz** [comment]
@wt-huang thank you for the response.
But what if I am not the one who created the gradient context (i.e., in Keras it is created internally) and I do not have access to the `tape` variable. How can I stop the gradient? 
Here is a toy example to illustrate the problem:

```
import tensorflow as tf

tfe = tf.contrib.eager
tf.enable_eager_execution()

a = tfe.Variable(0.)

def custom_loss(y_true, y_pred)
  #here I want to stop the gradient how can I do so?
  #I can only write the code of the custom loss function

def keras_backprop(loss, model, lr, x, y_true):
  with tf.GradientTape() as tape:
    y_pred = model(x)
    l = loss(y_true, y_pred)
  dw = tape.gradient(l, model.w)
  model.w = model.w - lr * dw
  return w
```

**Comment by nairouz** [comment]
Actually, I have found a workaround. I can simply wrap the tensor which I want to exclude from the gradient computation with `tfe.Variable()`.
The code would look something like this

def virtual_adversarial_loss(X, DAE_encoder):
    r_vadv = tfe.Variable(generate_virtual_adversarial_perturbation(DAE_encoder, X))
    p = tfe.Variable(DAE_encoder(X))
    q = model(X+r_vadv)
    loss = tf.reduce_mean(kl(p, q), axis=0)
    return tf.identity(loss, name="vat_loss")

But obviously, this is not intuitive at all.

**Comment by nairouz** [comment]
```import tensorflow as tftfe = tf.contrib.eagertf.enable_eager_execution()a = tfe.Variable(0.)def custom_loss(y_true, y_pred)  #here I want to stop the gradient how can I do so?  #I can only write the code of the custom loss functiondef keras_backprop(loss, model, lr, x, y_true):  with tf.GradientTape() as tape:    y_pred = model(x)    l = loss(y_true, y_pred)  dw = tape.gradient(l, model.w)  model.w = model.w - lr * dw  return w```

**Comment by wt-huang** [comment]
@nairouz Glad you found a workaround, will close this for now. Feel free to open a new issue if problem persists.

**Comment by nairouz** [comment]
Cc: @fchollet 
I think this should be fixed since there is no way to handle the `tape` in Keras eager execution mode :(
